rfd/0098-securing-docker-images.md

```diff
@@ -0,0 +1,181 @@
+fixes, etc. included) while still validating the provenance of the base image
+itself.
+
+It is technically possible for the image to be poisoned post-validation (e.g.
+in a shared build environment, where a malicious peer could re-tag a malicious
+image as the base).
+
+While we _could_ verify that the validated base image appears in the parent 
+chain of the final image, this is still no protection against the malicious 
+image being based on the same parent. We could _also_ assert that the final 
+image's parent chain has an expecetd number of steps (inferred from the number
+of steps in the Docker file), but this is erorr-prone and would drastically 
+reduce the flexibility of the build system.
+
+For the above reasons, Teleport images for public consumption must not be 
+built in such a shared environment.
+
+### Building the image
+
+The image will be built from a multi-stage docker file, using build stages to download
+and unpack the required debian packages and copy them into place on the distroless 
+image. 
+
+An example Dockerfile, assuming the Teleport Debian package is supplied by the 
+CI system, might look like something like:
+
+```Docker
+FROM debian:11 as dumb-init
+RUN apt update && apt-get download dumb-init && dpkg-deb -R dumb-init*.deb /opt/dumb-init
+
+FROM debian:11 as teleport
+COPY teleport*.deb
+RUN dpkg-deb -R teleport*.deb /opt/teleport
+
+FROM gcr.io/distroless/cc-debian11
+COPY --from=dumb-init /opt/dumb-init/bin/dumb-init /bin
+COPY --from=teleport /opt/teleport/bin/* /bin
+ENTRYPOINT ["/bin/dumb-init", "teleport", "start", "-c", "/etc/teleport/teleport.yaml"]
+```
+
+### Troubleshooting Images
+
+Troubleshooting a distroless image is hard, as there are no tools baked into
+the image to aid in debugging a deployment.
+
+If we need to add tooling in order to aid troubleshooting, I propose that we
+add a `teleport-debug` image, that builds on the distroless image to include
+things like a shell, tools from `busybox`, and so on. I would also suggest 
+that, while we should take as much care as possible when constructing and 
+monitoring this image, use of the debug image is considered "at your own risk".
+
+### Compatability Guarantees
+
+We have clients relying on the existing behaviour (and contents) of our images. We
+should treat releaseing these distroless images as a compatability break, and make 
+our customers aware of our intentions well in advance so that they can prepare.
+
+## 2. Scanning and monitoring the image
+
+There are many options for scanning and monitoring, but given we are already using the
+Amazon ECR, it seems most logical to use the built-in ECR scanning tools to detect
+known vulnerabilities in the final images.
+
+We are, in fact, already doing this on image upload, although I don't know if
+the signal from the scan is monitored in any way.
+
+I propose that we use the ECR Advanced Scanning tools, that can be set to run at
```

Comment:
How does this compare with other image scanning tools like Snyk? It looks like Snyk has support for [Github Actions](https://github.com/snyk/actions) which we would in theory integrate nicely with our CD, for example it could run on cron schedule and automatically update images if needed.

Reply:
Snyk is definitely a decent solution, but I was thinking ECR scanning would be a better initial fit for us.  The primary reason for that being we already have some ECR scanning and alerting in place that we can use as a starting place here.

Reply:
Adding Trivy sounds like a good option to me.  But it also will need a built image before it can be scanned